File: Payhub - SDK/ruby_SDK/PayHubSDK/lib/PayHubSDK/com/payhub/ws/util/ws_connections.rb
require 'uri'
require 'net/http'
require 'openssl'
require 'json'

class WsConnections
  # this is a set
  attr_writer :token

  def initialize(token=nil)
    @token=token
  end

  def setHeadersPost(operationUrl, token)

    url = URI(operationUrl)
    @token=token
    http = Net::HTTP.new(url.host, url.port)
    http.use_ssl = true
    http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    request = Net::HTTP::Post.new(url)
    request["content-type"] = 'application/json'
    request["accept"] = 'application/json'
    request["authorization"] = 'Bearer '+@token
    return http,request
  end

  def setHeadersGet(operationUrl, token)
    url = URI(operationUrl)
    @token=token
    http = Net::HTTP.new(url.host, url.port)
    http.use_ssl = true
    http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    request = Net::HTTP::Get.new(url)
    request["content-type"] = 'application/json'
    request["accept"] = 'application/json'
    request["authorization"] = 'Bearer '+@token
    return http,request
  end
  def setHeadersPut(operationUrl, token)
    url = URI(operationUrl)
    @token=token
    http = Net::HTTP.new(url.host, url.port)
    http.use_ssl = true
    http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    request = Net::HTTP::Put.new(url)
    request["content-type"] = 'application/json'
    request["accept"] = 'application/json'
    request["authorization"] = 'Bearer '+@token
    return http,request
  end
  def setHeadersPatch(operationUrl, token)
    url = URI(operationUrl)
    @token=token
    http = Net::HTTP.new(url.host, url.port)
    http.use_ssl = true
    http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    request = Net::HTTP::Patch.new(url)
    request["content-type"] = 'application/json'
    request["accept"] = 'application/json'
    request["authorization"] = 'Bearer '+@token
    return http,request
  end
  def doPost(http,request)
    response = http.request(request)
    case response.code.to_i
      when 200..399
        result = doGet(response['location'],@token)
        return result;
      else
        return generateErrorResponse(response)
    end
  end

  def doGet(url,token)
    http,request=setHeadersGet(url,token)
    response = http.request(request)
    case response.code.to_i
      when 200..399
        return response.read_body
      else
        return generateErrorResponse(response)
    end
  end

# @param [Object] http
# @param [Object] request
  def doPut(http,request)
    response = http.request(request)
    case response.code.to_i
      when 200..399
        return response
      else
        return generateErrorResponse(response)
    end
  end

  def doPatch(http,request)
    response = http.request(request)
    case response.code.to_i
      when 200..399
        return true
      else
        return generateErrorResponse(response)
    end
  end
  def doPostForRoles(http,request)
    response = http.request(request)
    case response.code.to_i
      when 200..399
        return response
      else
        return generateErrorResponse(response)
    end
  end
  # @param [Object] http
  # @param [Object] request
  # @param [Object] parameters
  def findTransactionReports(http, request, parameters)
    json = parameters.serialize_to_json #tojson
    request.body=json
    response = http.request(request)
    case response.code.to_i
      when 200..399
        return response.read_body
      else
        return generateErrorResponse(response)
    end
  end


  def generateErrorResponse(response)
    response2 = JSON.parse(response.read_body)
    if response2.include?('errors')
      return response.read_body
    elsif response2.to_s.empty?
      errors ||= Array.new
      error=Errors.new
      error.status="BAD_REQUEST"
      error.code="9995"
      error.severity="ERROR"
      error.location="404 no results"
      error.reason="No results for query"

      error = {"status"=>error.status,"code"=>error.code,"severity"=>error.severity,"location"=>error.location,"reason"=>error.reason}
      errors.push(error)
      informationToSend = {"errors"=>errors}
      result  = JSON.generate(informationToSend)
      return result.gsub('\\','').gsub('"{','{').gsub('}"','}')
    else
      errors ||= Array.new
      error=Errors.new
      error.status="BAD_REQUEST"
      error.code="9995"
      error.severity="ERROR"
      if response2.include?('cause') or response2.include?('message')
        error.location=response2['cause']
        error.reason=response2['message']
      end

      if response2.include?('error') or  response2.include?('error_description')
        error.location=response2['error']
        error.reason=response2['error_description']
      end
      error = {"status"=>error.status,"code"=>error.code,"severity"=>error.severity,"location"=>error.location,"reason"=>error.reason}
      errors.push(error)
      informationToSend = {"errors"=>errors}
      result  = JSON.generate(informationToSend)
      return result.gsub('\\','').gsub('"{','{').gsub('}"','}')
    end
  end
  def doPatchForBatch(http,request)
    response = http.request(request)
    case response.code.to_i
      when 200..399
        return response.read_body
      else
        return generateErrorResponse(response)
    end
  end
end

File: Payhub - SDK/php_SDK/com/payhub/ws/util/WsConnections.php
<?php

/**
 * Created by PhpStorm.
 * User: agustin
 * Date: 23/07/2015
 * Time: 13:04
 */
class WsConnections
{
    private $token;

    /**
     * WsConnections constructor.
     * @param $token
     */

    public function __construct($token)
    {
        if(!is_null($token)) {
            $this->token = $token;
        }
        else{
            $this->token="";
        }
    }

    /**
     * @return string
     */
    public function getToken()
    {
        return $this->token;
    }

    /**
     * @param string $token
     */
    public function setToken($token)
    {
        $this->token = $token;
    }


    public function setHeadersPost($WsURL,$token)
    {
        $this->token=$token;
        $request = curl_init();
        curl_setopt($request, CURLOPT_URL, $WsURL);
        curl_setopt($request, CURLOPT_HEADER, 1);
        curl_setopt($request, CURLOPT_CUSTOMREQUEST, "POST");
        curl_setopt($request, CURLOPT_HTTPHEADER, array(
                'Content-Type: application/json',
                'Accept: application/json',
                'Authorization: Bearer '. $this->token)
        );
        curl_setopt($request, CURLOPT_FOLLOWLOCATION, 1);
        curl_setopt($request, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($request, CURLOPT_SSL_VERIFYPEER, 0);
        return $request;
    }

    public function setHeadersGet($WsURL,$token)
    {
        $this->token=$token;
        $request = curl_init();
        curl_setopt($request, CURLOPT_URL, $WsURL);
       // curl_setopt($request, CURLOPT_HEADER, 1);
        curl_setopt($request, CURLOPT_CUSTOMREQUEST, "GET");
        curl_setopt($request, CURLOPT_HTTPHEADER, array(
                'Content-Type: application/json',
                'Accept: application/json',
                'Authorization: Bearer '. $this->token)
        );
     //   curl_setopt($request, CURLOPT_FOLLOWLOCATION, 1);
        curl_setopt($request, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($request, CURLOPT_SSL_VERIFYPEER, 0);
        return $request;
    }

    public function setHeadersPut($WsURL,$token)
    {
        echo $WsURL;

        $this->token=$token;
        $request = curl_init();
        curl_setopt($request, CURLOPT_URL, $WsURL);
        curl_setopt($request, CURLOPT_CUSTOMREQUEST, "PUT");
        curl_setopt($request, CURLOPT_HTTPHEADER, array(
                'Content-Type: application/json',
                'Accept: application/json',
                'Authorization: Bearer '. $this->token)
        );
        curl_setopt($request, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($request, CURLOPT_HEADER, 1);
        curl_setopt($request, CURLOPT_SSL_VERIFYPEER, 0);

        return $request;
    }

    public function setHeadersPatch($WsURL,$token)
    {
        $this->token=$token;
        $request = curl_init();
        curl_setopt($request, CURLOPT_URL, $WsURL);
        curl_setopt($request, CURLOPT_CUSTOMREQUEST, "PATCH");
        curl_setopt($request, CURLOPT_HTTPHEADER, array(
                'Content-Type: application/json',
                'Accept: application/json',
                'Authorization: Bearer '. $this->token)
        );
        curl_setopt($request, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($request, CURLOPT_HEADER, 1);
        curl_setopt($request, CURLOPT_SSL_VERIFYPEER, 0);

        return $request;
    }

    public function doPost($request, $_url){
        $response = curl_exec($request);
        $curl_info = curl_getinfo($request);
        $httpcode = curl_getinfo($request, CURLINFO_HTTP_CODE);
        curl_close($request);
        if ($httpcode==201){
            //find the url of the sale (Location header in the response)
//            $header = substr($response, 0, $header_size);
            $headers = substr($response, 0, $curl_info["header_size"]);
            preg_match("!\r\n(?:Location): *(.*?) *\r\n!", $headers, $matches);
            //$url contains the URL to GET the data from the Web Service
            $url = $matches[1];
            $request = $this->setHeadersGet($url,$this->getToken());
            $array = $this->doGet($request);
            return $array;
        }
        //There was an error with the WS
        else {
            $p = strpos($response, "\r\n\r\n");
            if( $p !== false ) {
                $rawBody = substr($response, $p + 4);
            }
            $data = json_decode($rawBody, true);
            if ($httpcode == 401) {//this is an token error
                return Errors::fromArrayForUnautenticated($data);
            }elseif($httpcode == 400){
                return Errors::fromArrayForBadRequest($data);
            }else{
                return $data;
            }
        }
    }
    public function doGet($request){
        $response = curl_exec($request);
        $httpcode = curl_getinfo($request, CURLINFO_HTTP_CODE);
        curl_close($request);
        $data = json_decode($response, true);
        if ($httpcode>=200 && $httpcode< 400){
            return $data;
        } else {
            if ($httpcode == 401) {
                $data = json_decode($response, true);
                return Errors::fromArrayForUnautenticated($data);
            }else{
                $rawBody="{\"errors\":[{\"status\":\"BAD_REQUEST\",\"code\": \"9995\",\"location\": \"404 Page not found.\",\"reason\": \"404 There aren't results for the query\",\"severity\": \"ERROR\"}]}";
                $data = json_decode($rawBody, true);
                return $data;
            }
        }

    }
    public function doPut($request,$json){
        echo $json;
        curl_setopt($request, CURLOPT_POSTFIELDS, $json);
        $response = curl_exec($request);
        $p = strpos($response, "\r\n\r\n");

        if( $p !== false ) {
            $rawBody = substr($response, $p + 4);
        }
        $data = json_decode($rawBody, true);
        return $data;
    }

    public function doPatch($request,$json){
        curl_setopt($request, CURLOPT_POSTFIELDS, $json);
        $response = curl_exec($request);
        $httpcode = curl_getinfo($request, CURLINFO_HTTP_CODE);
        if ($httpcode>=200 && $httpcode< 400){
            return true;
        } else {
            $p = strpos($response, "\r\n\r\n");
            $rawBody="";
            if( $p !== false ) {
                $rawBody = substr($response, $p + 4);
            }
            if ($httpcode == 401) {
                $data = json_decode($rawBody, true);
                return Errors::fromArrayForUnautenticated($data);
            }else{
                if($rawBody==""){
                    $rawBody="{\"errors\":[{\"status\":\"BAD_REQUEST\",\"code\": \"9995\",\"location\": \"404 Page not found.\",\"reason\": \"404 There aren't results for the query\",\"severity\": \"ERROR\"}]}";
                }
                $data = json_decode($rawBody, true);
                return $data;
            }
        }

    }
    public function doPostForRoles($request,$json){
        curl_setopt($request, CURLOPT_POSTFIELDS, $json);

        $response = curl_exec($request);
        $header_size = curl_getinfo($request, CURLINFO_HEADER_SIZE);
        $header = substr($response, 0, $header_size);
        $httpcode = curl_getinfo($request, CURLINFO_HTTP_CODE);
        $body = substr($response, $header_size);
        $data = json_decode($body, true);

//        var_dump($httpcode);
//        var_dump($data);
        if ($httpcode>=200 && $httpcode< 400){
            return $data;
        }elseif ($httpcode == 401) {//this is an token error
            return Errors::fromArrayForUnautenticated($data);
        }else{
            return $data;
        }
        return $data;

    }
    public function findTransactionReports($request,$json){
        $json=preg_replace('/,\s*"[^"]+":null|"[^"]+":null,?/', '', $json);
        curl_setopt($request, CURLOPT_POSTFIELDS, $json);
        $response = curl_exec($request);
        $httpcode = curl_getinfo($request, CURLINFO_HTTP_CODE);
        curl_close($request);
        $p = strpos($response, "\r\n\r\n");
        if( $p !== false ) {
            $rawHeades = substr($response, 0, $p);
            $rawBody = substr($response, $p + 4);
        }
        $data = json_decode($rawBody, true);
        if ($httpcode>=200 && $httpcode< 400){
            return $data;
        }elseif ($httpcode == 401) {//this is an token error
            return Errors::fromArrayForUnautenticated($data);
        }elseif($httpcode == 400){
            return Errors::fromArrayForBadRequest($data);
        }else{
            return $data;
        }

    }
}